### ISSUE TITLE
Nonexistent starting directory

### ISSUE THREAD

**Issue by BatmanAoD** [alacritty/alacritty#4995]
When alacritty is opened with a nonexistent starting directory, it panics:

```
---------------------------
Alacritty: Runtime Error
---------------------------
panicked at 'Unable to spawn shell: The directory name is invalid. (os error 267)', alacritty_terminal\src\tty\windows\conpty.rs:174:5

Press Ctrl-C to Copy
---------------------------
OK   
---------------------------
```

This makes sense, but it's inconvenient for WSL shells. When opening a WSL terminal instance with Alacritty when the WSL filesystem is not yet mounted (e.g. if WSL has not been launched since the host system last booted), if the configured starting directory is in the WSL filesystem, it appears there's no way to launch Alacritty without starting WSL another way first (e.g. using the Start menu entry for WSL).

It would be nice if the shell could simply fallback to the default (the process's current directory) rather than panicking; this way, the shell instance would still be usable, and, in the WSL case, the next shell instance to be opened will work as expected.

Another approach for the specific WSL issue could be to do some sort of check to see if the configured starting directory is in the WSL mount, and if so, wait until WSL is actually running before attempting to set the directory. This would result in a nicer user-experience for WSL, but I expect it would not be worth the complication on the implementation side.

### System

OS: Windows 10.0.19043
Version: alacritty 0.7.2 (5ac8060)

**Comment by BatmanAoD** [comment]
Sorry if this is a duplicate -- it seems like a pretty easy problem to run into, but I couldn't find any existing issues about the current behavior when the configured directory doesn't exist.

**Comment by chrisduerr** [comment]
Why set a starting directory if it doesn't exist yet?

**Comment by nixpulvis** [comment]
> Why set a starting directory if it doesn't exist yet?

Fair point, but I could see just starting in `$HOME`, `/`, or even just inheriting the CWD, as suggested, in the case this isn't set. Wouldn't that be better than a panic?

> Another approach for the specific WSL issue could be to do some sort of check to see if the configured starting directory is in the WSL mount, and if so, wait until WSL is actually running before attempting to set the directory. This would result in a nicer user-experience for WSL, but I expect it would not be worth the complication on the implementation side.

This is not something Alacritty is concerned with.

**Comment by chrisduerr** [comment]
Well if you use `--working-directory`, Alacritty will emit an error rather than crashing. But for a configuration option I'd expect the path specified to exist.

**Comment by BatmanAoD** [comment]
@chrisduerr 
 
> Why set a starting directory if it doesn't exist yet?

As I explained, WSL directories aren't accessible if the WSL filesystem hasn't been mounted yet, but the filesystem is auto-mounted in the process of launching WSL, so even if the directory isn't available when the shell starts, it will become available almost immediately and remain accessible until the filesystem is unmounted again. (In my experience, I've only seen host reboots unmount the filesystem; shutting down WSL before launching Alacritty does not trigger the error I described.)

> Well if you use `--working-directory`, Alacritty will emit an error rather than crashing.

....ah, I thought I remembered having this working at one point; I must have configured my Windows shortcut for the Alacritty WSL terminal to use this option, and left the configured starting directory alone. Thanks.

Why is there a difference between this command-line option and the configuration option? It seems preferable to me to emit errors for broken configuration, but still launch a usable terminal instance, than to crash.

**Comment by chrisduerr** [comment]
> Why is there a difference between this command-line option and the configuration option?

Because a path passed through the commandline is much more ephemeral. A path permanently configured as your default working directory in your configuration file should always exist.

**Comment by amitbha** [comment]
give an option to determine whether fallback to $HOME when nonexist?

**Comment by BatmanAoD** [comment]
@amitbha In the WSL case, `$HOME` _is_ the directory that doesn't exist yet, because the whole guest filesystem isn't mounted yet. The weirdest part, to me, is that `$HOME/.bashrc` is always successfully sourced, so it's not like the shell is going to be interactive before the filesystem is actually available. Crashing in this case just does not make any sense from a user perspective.

**Comment by BatmanAoD** [comment]
Even a brief delay-with-retry (on the order of milliseconds) would work, I'd expect.

**Comment by BatmanAoD** [comment]
I think I may have found a better solution. Launching wsl using `wsl.exe` instead of `bash.exe` lets you specify that the shell should start in the home directory, and presumably it will also respect a non-default shell configured within WSL itself.

```
shell:
  program: wsl.exe
  args:
    - --cd ~
```

CC @amitbha

**Comment by amitbha** [comment]
> I think I may have found a better solution. Launching wsl using `wsl.exe` instead of `bash.exe` lets you specify that the shell should start in the home directory, and presumably it will also respect a non-default shell configured within WSL itself.
> 
> ```
> shell:
>   program: wsl.exe
>   args:
>     - --cd ~
> ```
> 
@chrisduerr

**Comment by YAMLcase** [comment]
> I think I may have found a better solution. Launching wsl using `wsl.exe` instead of `bash.exe` lets you specify that the shell should start in the home directory, and presumably it will also respect a non-default shell configured within WSL itself.

Came here looking for this. Thanks

**Comment by nixpulvis** [comment]
Doesn #8457 address this?

**Comment by chrisduerr** [comment]
Master works fine with an invalid directory in the config.
